### ISSUE TITLE
3D UNet error: blob size exceeds INT_MAX

### ISSUE THREAD

**Issue by pr4deepr** [lmb-freiburg/Unet-Segmentation#59]
Hi
  I am trying to use finetune a model for 3D cell segmentation using the existing 3D model in the package: 3d_cell_net_microspores-fluorescence_v1. I keep getting this error: "

**_fiji unet error F0213 01:12:15.345480  4312 blob.cpp:34] Check failed: shape[i] <= 0x7fffffff / count_ (232 vs. 229) blob size exceeds INT_MAX_**

What does this mean and how can I rectify this? From what I gather, the HDF5 filesize is over a certain limit?

The cells in my images have been annotated every second slice. My original image was 1024x1024, which I resized to 512x512.  I have three training images with two of them have around 30 slices and the third one has 20 slices.  One image stack was used for validation. When finetuning, I used a tileshape of : 256x256x140.

I tried splitting the original 1024x1024 into 4 stacks of 256x256 each, with tileshape of 252x252x124 and still got a similar error:
 **_F0213 06:22:41.204368 10213 blob.cpp:34] Check failed: shape[i] <= 0x7fffffff / count_ (248 vs. 242) blob size exceeds INT_MAX_**

The backend is on a virtual machine on a cloud service and has been setup using pre-built binaries according to instructions on the lmb website.  Config: Ubuntu 16.10, NVIDIA Tesla T4 15 GB. 

I have used UNet segmentation successfully for 2D images on this setup.

Thanks
Pradeep

**Comment by ThorstenFalk** [comment]
Do you have cuDNN enabled? BLOB_SIZE exceeds INT_MAX should only happen if caffe wants to use im2col which allocates incredibly huge intermediate blobs.

**Comment by pr4deepr** [comment]
Thanks @ThorstenFalk . I think it was the cuDNN. 
I set everything up from source on a VM with Ubuntu 18.04, CUDA toolkit and cuDnn installed. 
Followed the instructions on your website under "**Setup on own server (from source**)"
I tried splitting the original 1024x1024 stacks into 4 stacks of 256x256 each, so 12 stacks in total. finetuning with tileshape of 252x252x124. Estimates memory of 12616MB memory with cuDNN
The graphs for IOU, F1 appear, but now I get this error:
I0214 06:06:46.863917 19404 solver.cpp:414]     Test net output #3: loss_valid = 0.038803 (* 1 = 0.038803 loss)
F0214 06:06:48.374413 19404 benchmark.cpp:92] Check failed: error == cudaSuccess (716 vs. 0)  misaligned address
*** Check failure stack trace: ***
    @     0x7f64d07230cd  google::LogMessage::Fail()
    @     0x7f64d0724f33  google::LogMessage::SendToLog()
    @     0x7f64d0722c28  google::LogMessage::Flush()
    @     0x7f64d0725999  google::LogMessageFatal::~LogMessageFatal()
    @     0x7f64d0e91142  caffe::Timer::MilliSeconds()
    @     0x7f64d0e906ca  caffe::Timer::Seconds()
    @     0x7f64d0e7444a  caffe::Solver<>::Step()
    @     0x7f64d0e74c6d  caffe::Solver<>::Solve()
    @     0x55a71ef55e73  train()
    @     0x55a71ef52161  main
    @     0x7f64cf4aab97  __libc_start_main
    @     0x55a71ef52cea  _start

Thanks heaps.

**Comment by ThorstenFalk** [comment]
12616MB is a rough estimate, but should be sufficiently far away from your 15GB limit, but I have no experience with cards with more than 12GB. Could you try with slightly smaller tiles, just to be sure that you did not accidentally ran out of memory?

**Comment by pr4deepr** [comment]
Hi
  I tried with one image for finetuning (no validation images), tilesize of 188*188*92, it was the least I could go. Plugin predicted 4820 MB usage and It crashed with a similar error: 
16140 math_functions.hpp:203] Check failed: error == cudaSuccess (716 vs. 0)  misaligned address
The card reached a usage of 5682MB before crashing.

I thought I'd test out **segmentation only** with the stacks using the default 3D fluorescence model to see if it crashed. I tried different combination of tilesizes and monitored the card using the nvidia-smi command; 

- 84x84x28 (predicted :4702 and actual: 5625MB)
- 100x100x28 (predicted :5280 and actual: 6221 MB)
- 132x132x44 (predicted :7634 and actual: 8633MB)
- 148x148x52 (predicted :9112 and actual: 10143MB)
- 164x164x52 (predicted :10120 and actual: 11091 MB)
These combinations worked and the plugin returned a segmented stack.

**But** when I tried a combination of 
180x180x52, which predicted 10407 MB usage, 
the plugin **crashed** with an error:
Max: F0215 03:28:36.875310 12656 cudnn_conv_layer.cpp:172] Check failed: status == CUDNN_STATUS_SUCCESS (3 vs. 0)  CUDNN_STATUS_BAD_PARAM
Most of the tile combinations that predicted VRAM usage of > 10120 MB RAM threw an error.

Not sure what is happening, but hoping this information helps..
Thanks
Pradeep

**Comment by ThorstenFalk** [comment]
Hmz, OK, the  estimates seem to be OK, slightly underestimated, that's what I mean with not precise. Depending on cuDNN there is some overhead the plugin cannot foresee.

So it's not about the memory per se. I was just curious because when using tile sizes that are close to the limit any kind of error can occur. I very often encountered CUDNN_STATUS_BAD_PARAM with caffe when reaching the memory limit although I was always sure it should fit although with tiny margin. The last update of caffe was before 15GB cards were around so maybe support beyond 12GB was never tested.

Your Tesla card is bleeding edge technology and supports cuda compute capability 7.5 which is way beyond the last compute capability when caffe ran out of updates. I would still try to build caffe_unet explicitly for CC 7.5 by extending the corresponding fields in the cmake call. Maybe this helps.

**Comment by pr4deepr** [comment]
Thanks for your replies. I did build caffe_unet explicitly for CC 7.5, but still got the error.
 
I started from scratch again with a Tesla P100 (16 GB), used Ubuntu 16.04 with CUDA 8 and cuDNN 7. This time, I used the prebuilt binary from your website: caffe_unet_package_16.04_gpu_cuDNN.zip. 
Using this setup I was able to run finetuning and segmentation without any errors and it was using around 15 GB of the GPU memory. It seems to be working now!!

I have a few other questions: 

I have a question about finetuning with 3D data. I understand this depends on the data, but how do you determine what the number of iterations should be? What are some things I should look out for and that is different from 2D data?

I would like to minimise the cost of running a VM, so whats the minimum amount of RAM, CPU configuration and storage you  would recommend?

Also, the prebuilt binary you distribute, are they the unet folder you built from source in a zip file?

**Comment by ThorstenFalk** [comment]
Finetuning 3D data is the same as finetuning with 2D data, the only difference is that you can only use small sub-volumes per iteration and therefore you will need many epochs until the network saw all your data. Additionally, each iteration requires more time than in the 2D case, therefore, the overall training times can become huge even with bleeding edge GPU hardware. I remember a training running for 6 months and the model still improved. Bad news is, during that training, validation loss was saturated already after few hours, so no hint from that side.

A lot of RAM helps. If you are able to keep your training set in RAM you avoid expensive File-IO and can reduce training times by up to 80%. So if you rent a machine, time reduction can help more than savings from cheaper hardware. The CPU actually doesn't matter too much. Very likely your bottleneck will be data access. So I would go for one of the smallest 4th generation Xeon for fast memory clock and a fast SSD.

Yes the pre-built binaries are simply a zipped version of our installation folder.

**Comment by pr4deepr** [comment]
Thanks for that. I appreciate your help in this matter. 

If I am using, say 3 stacks of 512x512x30 with tile sizes of  ~256x256x140, what are the minimum iterations you would recommend? For 2D finetuning, I remember seeing around a max of 1000 as per your supplementary info in the paper.

Thanks for the input on the VM. With RAM, I was asking about system RAM (not the GPU RAM), is that what you meant?

**Comment by ThorstenFalk** [comment]
You have stacks with 30 slices only? Why would you go for blobs with depth 140 then? I assume the voxels are anisotropic and you do only two subsampling steps in z-direction and three in x- and y-? Intuitively I'd say 140 slices for the input blobs are an overkill, I would at least half it and spend the saved memory in larger x-y dimensions.

Well, finetuning 1k iterations can yield very good results if the target domain is very similar to the source domain. With domain shift, e.g. more input channels, more output classes, detection instead of segmentation, other imaging modality, ... you may need many more. From my experience, segmentation tasks are relatively easy and results are quickly satisfactory with IoUs close to 0.9. Detection is a completely different story, I trained a detection model for 500k  iterations and it was still improving... I had to abort training when I started my new job, so I can't tell whether in the end it would have perfectly solved the problem.

From scratch, our 3D models were trained for up to 80k iterations, usually after 10k you see whether the model evolves in the desired direction and especially the segmentation model we ship was already saturated after 24k iterations.

The 2d cell net was trained for only 160k iterations, which was one day of training. The validation loss dropped dramatically afterwards, so when writing the paper, we thought we would run into overfitting with more than 160k. I repeated the experiment later and let it train for another 7 million iterations. After approx. a million iterations validation started to slowly recover and after approx. 2 million iterations it surpassed the old value and steadily increased. In the end the model was much better than the original cell net wrt. validation loss. The reason was the pollen dataset. The network learnt the concept of segmenting bright or dark particles from homogeneous background very fast, but the pollen dataset contained a lot of background clutter. After 160k iterations the network performed well on 90% of the data but the background clutter of the pollen class was quite often classified as cells. It took 2 million more iterations to distinguish background clutter from pollen which in the end made the model much better for the scenario of dirty samples.

So long story short: The amount of finetuning you need is completely data dependent. Usually I ended up with a scheme in which I finetuned for 2k, if results were better than after 1k I trained for 4k and compared to 2k, and continued doubling the finetuning time until the difference between x/2 iterations and x iterations was negligible (when looked at it with bad blur filter) both quantitatively and also qualitatively when looking at sample images.

Regarding your second question: Yes I meant system RAM, not GPU memory. The more RAM, the more data from disk can be cached and the higher the chance that data need not be fetched from hard drive. If you want to be 100% sure and your dataset allows, put it into a RAM disk.

**Comment by pr4deepr** [comment]
Thanks heaps for the advice. I am relatively new to this, so still getting my head around it . We setup the microscope so it determines the optimal z step sampling. Pretty sure the voxels are isotropic as in they are: 0.378x0.378x0.346 microns in the image. So, I am Assuming that if the steps in the z direction are low, can use a lower depth. It won't really go over 50 slices. 

I am not doing any detection tasks for now. It's mostly segmentation of cells in 3D stacks. Its relatively clean staining, but there are a lot of touching objects, so need to make sure the annotation is really good to help with distinguishing between them. Its interesting to hear how you've done the training, your approaches and insights into this. I like the idea of doubling the finetuning time and checking the difference between iterations. 

I'm putting some time into reading about using deep learning for image segmentation. Your plugin provides me a hands-on way of understanding it and applying it to my data.  I'm glad I don't have to do any training from scratch, yet! :)
Thanks again for your detailed explanation.
